### PR TITLE
thrift: support onLocalReply to inform filters of local replies

### PR DESCRIPTION
**Pull request by JuniorHsu**
Signed-off-by: kuochunghsu <kuochunghsu@pinterest.com>

Commit Message:
Similar setup with http onLocalReply in #15172

Additional Description:
Risk Level: low
Testing: unit test
Docs Changes: current.yaml
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/extensions/filters/network/thrift_proxy/filters/filter.h [hunk 1]
```diff
@@ -142,12 +142,44 @@ class EncoderFilterCallbacks : public virtual FilterCallbacks {
...
+/**
+ * Return codes for onLocalReply filter invocations.
+ */
+enum class LocalErrorStatus {
+  // Continue sending the local reply after onLocalError has been sent to all filters.
+  Continue,
+
+  // Continue sending onLocalReply to all filters, but reset the stream once all filters have been
+  // informed rather than sending the local reply.
+  ContinueAndResetStream,
```
Comment (zuercher) [inline source/extensions/filters/network/thrift_proxy/filters/filter.h:154]: Is the description correct? Seems like `parent_.sendLocalReply` is always called and this causes the stream to be reset after the reply.

Comment (JuniorHsu) [reply source/extensions/filters/network/thrift_proxy/filters/filter.h:154]: you're right. Let me see which makes sense, changing description or implementation.

**Comment by JuniorHsu** [comment]
After more deliberation, It would make more sense to let filter stop the response instead of ending stream.
However, currently we don't have use case at the moment.

To reduce risk, I'd like to remove the code for end stream handling for now,
but keep the response struct type for flexibility (i.e., if someone implement end_stream or reset_imminent, the filters still compile).

@zuercher would you mind to take another look? Thanks!

**Review by zuercher — COMMENTED** [review]
One small thing on the release notes, but otherwise looks good to me.

File: changelogs/current.yaml [hunk 2]
```diff
@@ -146,6 +146,7 @@ new_features:
     added support for access logging.
 - area: thrift
   change: |
+    added support onLocalReply to inform filters of local replies.
```
[the remark this answers is not in the document]
Comment (JuniorHsu) [reply changelogs/current.yaml]: Thanks for catching this! `git merge` accidentally brought this.

**Review by rgs1 — APPROVED** [review]
LGTM, thanks!

**Review by zuercher — APPROVED** [review]

**Pull request merged**